lib/msal-core/src/Utils.ts

```diff
@@ -567,48 +581,224 @@ export class Utils {
   //#region ExtraQueryParameters Processing (Extract?)
 
   /**
-   *
+   * returns 'true' is developer has some SSO info in the token request, else 'false'
    * @param extraQueryParameters
    */
-  static checkSSO(extraQueryParameters: string) {
-    return  !(extraQueryParameters &&  ((extraQueryParameters.indexOf(Constants.login_hint) !== -1 ||  extraQueryParameters.indexOf(Constants.sid) !== -1 )));
+  static checkSSO(request: AuthenticationParameters) {
+    return  (request.sid  || request.loginHint);
   }
 
-   /**
+  /**
    * Constructs extraQueryParameters to be sent to the server for the AuthenticationParameters set by the developer
    * in any login() or acquireToken() calls
-   *
    * @param idTokenObject
-   * @param login_hint
    * @param extraQueryParameters
+   * @param sid
+   * @param loginHint
    */
   //TODO: check how this behaves when domain_hint only is sent in extraparameters and idToken has no upn.
-  //TODO: Test all paths thoroughly
-  static constructUnifiedCacheExtraQueryParameter(idTokenObject: any, extraQueryParameters?: string) {
-    if (idTokenObject) {
+  static constructUnifiedCacheExtraQueryParameter(extraQueryParameters: string, account: User, sid: string, loginHint: string, idTokenObject: any): string {
+
+    // preference order: account > sid > login_hint
+    let ssoType;
+    let ssoData;
+
+    // if account info is passed, account.sid > account.login_hint
+    if (account) {
```

Comment:
+1. These changes are in progress. This API and addHintParameters are something that I wanted to further strip down to utils.

Reply:
I tried this. It doesn't really add much value w.r.t optimization as what we finally want is trickle down to each option as we find out that the initial option is not available.

Let me think of a more simpler way to do this instead of processing all options, may be send only the relevant options. I will get back on this later. Closing this now.